django/contrib/auth/decorators.py

```diff
@@ -35,6 +35,27 @@ def _wrapper_view(request, *args, **kwargs):
 
             return redirect_to_login(path, resolved_login_url, redirect_field_name)
 
+        if asyncio.iscoroutinefunction(view_func):
+
+            @wraps(view_func)
+            async def _wrapper_view(request, *args, **kwargs):
+                if asyncio.iscoroutinefunction(test_func):
+                    result = await test_func(await request.auser())
+                else:
+                    result = test_func(await request.auser())
+
+                if result:
+                    return await view_func(request, *args, **kwargs)
+                return _redirect_to_login(request)
+
+        else:
+
+            @wraps(view_func)
+            def _wrapper_view(request, *args, **kwargs):
+                if test_func(request.user):
+                    return view_func(request, *args, **kwargs)
+                return _redirect_to_login(request)
```

Comment:
Same here, I think we should allow an `async` `test_func` that wraps a sync function.

Reply:
Okay, thank you for your advice. I completely understand and make sense. I'll update the code soon.